File: quadtree/quadtree_cpp/src/QuadTree.cpp
#include "..\include\QuadTree.h"
#include <raylib.h>


QuadTree::QuadTree(Custom::Rectangle& bounds, int capacity) 
{
	this->bounds = bounds;
	this->capacity = capacity;
	
	points = new Custom::Point[capacity];

	point_counts = 0;
	divided = false;
}


void QuadTree::subdivide() 
{
	// Origin is top left
	float half_w = bounds.w / 2;
	float half_h = bounds.h / 2;

	Custom::Rectangle lt{ bounds.x, bounds.y, half_w, half_h };
	Custom::Rectangle lb{ bounds.x, bounds.y + half_h, half_w, half_h };
	Custom::Rectangle rt{ bounds.x + half_w, bounds.y, half_w, half_h };
	Custom::Rectangle rb{ bounds.x + half_w, bounds.y + half_h, half_w, half_h };

	childs[0] = new QuadTree(lb, capacity);
	childs[1] = new QuadTree(lt, capacity);
	childs[2] = new QuadTree(rb, capacity);
	childs[3] = new QuadTree(rt, capacity);

	divided = true;
}

// Check if current level contains point
bool QuadTree::contains(Custom::Point p) 
{
	if (p.x < bounds.x) return false;
	if (p.x > bounds.x + bounds.w) return false;
	if (p.y < bounds.y) return false;
	if (p.y > bounds.y + bounds.h) return false;
	return true;
}

// Returns whether insertion in current node was successful
bool QuadTree::insert(Custom::Point& point) 
{	
	// If does not contain point, stop and return false
	if (!contains(point)) 
		return false;

	// Can insert
	if (point_counts < capacity) {
		points[point_counts++] = point;
		return true;
	}
	// Can't insert, subdivide
	else {
		if (!divided) 
			subdivide();

		bool insertion = false;
		for (auto child : childs) {
			if (child->insert(point)) {
				insertion = true;
				break;
			}
		}
		return insertion;
	}
}

void QuadTree::draw() 
{
	// Draw the boundary
	DrawRectangleLines(bounds.x, bounds.y, bounds.w, bounds.h, WHITE);
	
	// Draw all the points in this boundary
	for (int i = 0; i < point_counts; i++)
		DrawCircle(points[i].x, points[i].y, 1, RED);
	
	// If children exist, draw them
	if (divided) {
		for (auto child : childs) 
			child->draw(); 
	}
}

File: quadtree/quadtree_py/quadtree.py
#%%
from pyray import *

# %%
from dataclasses import dataclass

@dataclass
class Rectangle:
    x: float
    y: float
    w: float
    h: float


@dataclass
class Point:
    x: float
    y: float


class QuadTree:
    def __init__(self, rectangle: Rectangle, capacity: int):
        
        self.boundary: Rectangle = rectangle
        self.capacity: int = capacity

        self.subdivided: bool = False
        self.points: list[Point] = []
        self.childs: list[Rectangle] = [None, None, None, None]
    

    def __contains__(self, point: Point) -> bool:
        """Returns True if the point is inside the rectangle"""
        if point.x < self.boundary.x:
            return False
        if point.y < self.boundary.y:
            return False
        if point.x > self.boundary.x + self.boundary.w:
            return False
        if point.y > self.boundary.y + self.boundary.h:
            return False
        return True
    

    def subdivide(self) -> None:
        self.subdivided = True

        half_w = self.boundary.w/2
        half_h = self.boundary.h/2
        
        # Left top
        self.childs[0] = QuadTree(Rectangle(x=self.boundary.x, y=self.boundary.y, w=half_w, h=half_h), self.capacity)
        # Left bottom
        self.childs[1] = QuadTree(Rectangle(x=self.boundary.x, y=self.boundary.y + half_h, w=half_w, h=half_h), self.capacity)
        # Right top
        self.childs[2] = QuadTree(Rectangle(x=self.boundary.x + half_w, y=self.boundary.y, w=half_w, h=half_h), self.capacity)
        # Right bottom
        self.childs[3] = QuadTree(Rectangle(x=self.boundary.x + half_w, y=self.boundary.y + half_h, w=half_w, h=half_h), self.capacity)

    
    def insert(self, point: Point) -> bool:
        """Inserts a point into the quadtree"""
        
        if point not in self:
            return False

        if len(self.points) < self.capacity:
            self.points.append(point)
            return True
        else:
            if not self.subdivided:
                self.subdivide()
            
            insertation = False
            for child in self.childs:
                if child.insert(point):
                    insertation = True
                    break
            return insertation
                
    
    def query(self, point):
        """Returns the rectangle that contains the point"""
        pass
    
    
    def draw(self):
        # draw_rectangle_lines(self.boundary.x, self.boundary.y, self.boundary.w, self.boundary.h, WHITE)
        draw_rectangle_lines(int(self.boundary.x), int(self.boundary.y), int(self.boundary.w), int(self.boundary.h), WHITE)

        for point in self.points:
            draw_circle(point.x, point.y, 2, RED)
        
        if self.subdivided:
            for child in self.childs:
                child.draw()

    
    
def main():
    init_window(800, 800, 'Quadtree')
    
    qtree = QuadTree(Rectangle(0, 0, 800, 800), 4)
    
    while not window_should_close():
        begin_drawing()
        clear_background(BLACK)
        
        qtree.insert(Point(get_mouse_x(), get_mouse_y()))

        qtree.draw()
        
        end_drawing()
        
    close_window()


if __name__ == "__main__":
    main()
    
    